File: Drug-stock predictor/Desktop/Drug-stock-predictor/med-predict/go-backend/internal/handlers/stock.go
package handlers

import (
	"database/sql"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"med-predict/go-backend/internal/middleware"
	"med-predict/go-backend/internal/models"
)

type StockHandler struct {
	db *sql.DB
}

// NewStockHandler creates a new stock handler
func NewStockHandler(db *sql.DB) *StockHandler {
	return &StockHandler{db: db}
}

// ListMedicines lists all medicines for a pharmacy
func (h *StockHandler) ListMedicines(c *gin.Context) {
	_, pharmacyID, _, err := middleware.GetUserFromContext(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"detail": "User not authenticated"})
		return
	}

	rows, err := h.db.Query(`
		SELECT id, pharmacy_id, name, generic_name, dosage, unit, quantity_on_hand,
		       reorder_level, unit_cost, selling_price, expiry_date, manufacturer_id,
		       batch_number, is_active, created_at, updated_at
		FROM medicines
		WHERE pharmacy_id = $1 AND is_active = true
		ORDER BY name
	`, pharmacyID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "Failed to fetch medicines"})
		return
	}
	defer rows.Close()

	medicines := []models.MedicineResponse{}
	for rows.Next() {
		var med models.MedicineResponse
		err := rows.Scan(
			&med.ID, &med.PharmacyID, &med.Name, &med.GenericName, &med.Dosage, &med.Unit,
			&med.QuantityOnHand, &med.ReorderLevel, &med.UnitCost, &med.SellingPrice,
			&med.ExpiryDate, &med.ManufacturerID, &med.BatchNumber, &med.IsActive,
			&med.CreatedAt, &med.UpdatedAt,
		)
		if err != nil {
			continue
		}
		medicines = append(medicines, med)
	}

	c.JSON(http.StatusOK, medicines)
}

// CreateMedicine creates a new medicine
func (h *StockHandler) CreateMedicine(c *gin.Context) {
	_, pharmacyID, _, err := middleware.GetUserFromContext(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"detail": "User not authenticated"})
		return
	}

	var req models.MedicineCreate
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
		return
	}

	medicineID := generateID("med")
	now := time.Now()

	query := `
		INSERT INTO medicines (id, pharmacy_id, name, generic_name, dosage, unit, quantity_on_hand,
		                        reorder_level, unit_cost, selling_price, expiry_date, manufacturer_id,
		                        batch_number, is_active, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16)
	`

	_, err = h.db.Exec(query, medicineID, pharmacyID, req.Name, req.GenericName, req.Dosage, req.Unit,
		req.QuantityOnHand, req.ReorderLevel, req.UnitCost, req.SellingPrice, req.ExpiryDate,
		req.ManufacturerID, req.BatchNumber, true, now, now)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "Failed to create medicine"})
		return
	}

	c.JSON(http.StatusCreated, gin.H{"id": medicineID, "message": "Medicine created successfully"})
}

// UpdateMedicine updates a medicine
func (h *StockHandler) UpdateMedicine(c *gin.Context) {
	_, pharmacyID, _, err := middleware.GetUserFromContext(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"detail": "User not authenticated"})
		return
	}

	medicineID := c.Param("id")

	// Check if medicine belongs to user's pharmacy
	var existingPharmacyID string
	err = h.db.QueryRow("SELECT pharmacy_id FROM medicines WHERE id = $1", medicineID).Scan(&existingPharmacyID)
	if err != nil || existingPharmacyID != pharmacyID {
		c.JSON(http.StatusForbidden, gin.H{"detail": "Access denied"})
		return
	}

	var req models.MedicineUpdate
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
		return
	}

	// Build dynamic update query
	updates := "updated_at = $1"
	args := []interface{}{time.Now()}
	argIndex := 2

	if req.Name != nil {
		updates += ", name = $" + strconv.Itoa(argIndex)
		args = append(args, *req.Name)
		argIndex++
	}
	if req.QuantityOnHand != nil {
		updates += ", quantity_on_hand = $" + strconv.Itoa(argIndex)
		args = append(args, *req.QuantityOnHand)
		argIndex++
	}
	// Add other fields as needed...

	args = append(args, medicineID)

	query := `UPDATE medicines SET ` + updates + ` WHERE id = $` + strconv.Itoa(argIndex)
	_, err = h.db.Exec(query, args...)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "Failed to update medicine"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Medicine updated successfully"})
}

// DeleteMedicine soft deletes a medicine
func (h *StockHandler) DeleteMedicine(c *gin.Context) {
	_, pharmacyID, _, err := middleware.GetUserFromContext(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"detail": "User not authenticated"})
		return
	}

	medicineID := c.Param("id")

	// Check if medicine belongs to user's pharmacy
	var existingPharmacyID string
	err = h.db.QueryRow("SELECT pharmacy_id FROM medicines WHERE id = $1", medicineID).Scan(&existingPharmacyID)
	if err != nil || existingPharmacyID != pharmacyID {
		c.JSON(http.StatusForbidden, gin.H{"detail": "Access denied"})
		return
	}

	query := `UPDATE medicines SET is_active = false, updated_at = $1 WHERE id = $2`
	_, err = h.db.Exec(query, time.Now(), medicineID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "Failed to delete medicine"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Medicine deleted successfully"})
}

// GetMedicine retrieves a specific medicine
func (h *StockHandler) GetMedicine(c *gin.Context) {
	_, pharmacyID, _, err := middleware.GetUserFromContext(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"detail": "User not authenticated"})
		return
	}

	medicineID := c.Param("id")

	var med models.MedicineResponse
	query := `
		SELECT id, pharmacy_id, name, generic_name, dosage, unit, quantity_on_hand,
		       reorder_level, unit_cost, selling_price, expiry_date, manufacturer_id,
		       batch_number, is_active, created_at, updated_at
		FROM medicines
		WHERE id = $1 AND pharmacy_id = $2
	`

	err = h.db.QueryRow(query, medicineID, pharmacyID).Scan(
		&med.ID, &med.PharmacyID, &med.Name, &med.GenericName, &med.Dosage, &med.Unit,
		&med.QuantityOnHand, &med.ReorderLevel, &med.UnitCost, &med.SellingPrice,
		&med.ExpiryDate, &med.ManufacturerID, &med.BatchNumber, &med.IsActive,
		&med.CreatedAt, &med.UpdatedAt,
	)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "Medicine not found"})
		return
	}

	c.JSON(http.StatusOK, med)
}

File: Drug-stock predictor/Desktop/Drug-stock-predictor/med-predict/py-backend/app/handlers/stock.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.db import get_db
from app.models import (
    MedicineCreate,
    MedicineUpdate,
    MedicineResponse,
)
from app.models.database import Medicine
from app.middleware.auth import get_current_user
from app.models import TokenData
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/stock", tags=["stock"])


@router.get("/medicines", response_model=list[MedicineResponse])
async def list_medicines(
    current_user: TokenData = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List all medicines for the user's pharmacy"""
    medicines = db.query(Medicine).filter(
        Medicine.pharmacy_id == current_user.pharmacy_id
    ).all()
    
    response = []
    for medicine in medicines:
        # Calculate status
        now = datetime.now().date()
        status_val = "ok"
        if medicine.expiry_date < now:
            status_val = "expired"
        elif (medicine.expiry_date - now).days <= medicine.notification_days:
            status_val = "expiring"
        elif medicine.quantity_remaining <= medicine.reorder_level:
            status_val = "low"
        
        response.append(MedicineResponse(
            id=str(medicine.id),
            pharmacy_id=str(medicine.pharmacy_id),
            name=medicine.name,
            generic_name=medicine.generic_name,
            category=medicine.category,
            unit=medicine.unit,
            quantity_total=medicine.quantity_total,
            quantity_remaining=medicine.quantity_remaining,
            expiry_date=medicine.expiry_date.isoformat(),
            batch_number=medicine.batch_number,
            supplier=medicine.supplier,
            unit_cost=medicine.unit_cost,
            reorder_level=medicine.reorder_level,
            notification_days=medicine.notification_days,
            status=status_val,
            created_by=str(medicine.created_by),
            created_at=medicine.created_at,
            updated_at=medicine.updated_at,
        ))
    
    return response


@router.post("/medicines", response_model=MedicineResponse)
async def create_medicine(
    medicine_data: MedicineCreate,
    current_user: TokenData = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Create a new medicine entry"""
    try:
        expiry_date = datetime.fromisoformat(medicine_data.expiry_date).date()
        
        new_medicine = Medicine(
            pharmacy_id=current_user.pharmacy_id,
            name=medicine_data.name,
            generic_name=medicine_data.generic_name,
            category=medicine_data.category,
            unit=medicine_data.unit,
            quantity_total=medicine_data.quantity_total,
            quantity_remaining=medicine_data.quantity_remaining,
            expiry_date=expiry_date,
            batch_number=medicine_data.batch_number,
            supplier=medicine_data.supplier,
            unit_cost=medicine_data.unit_cost,
            reorder_level=medicine_data.reorder_level,
            notification_days=medicine_data.notification_days,
            created_by=current_user.user_id,
        )
        
        db.add(new_medicine)
        db.commit()
        db.refresh(new_medicine)
        
        # Calculate status
        now = datetime.now().date()
        status_val = "ok"
        if new_medicine.expiry_date < now:
            status_val = "expired"
        elif (new_medicine.expiry_date - now).days <= new_medicine.notification_days:
            status_val = "expiring"
        elif new_medicine.quantity_remaining <= new_medicine.reorder_level:
            status_val = "low"
        
        return MedicineResponse(
            id=str(new_medicine.id),
            pharmacy_id=str(new_medicine.pharmacy_id),
            name=new_medicine.name,
            generic_name=new_medicine.generic_name,
            category=new_medicine.category,
            unit=new_medicine.unit,
            quantity_total=new_medicine.quantity_total,
            quantity_remaining=new_medicine.quantity_remaining,
            expiry_date=new_medicine.expiry_date.isoformat(),
            batch_number=new_medicine.batch_number,
            supplier=new_medicine.supplier,
            unit_cost=new_medicine.unit_cost,
            reorder_level=new_medicine.reorder_level,
            notification_days=new_medicine.notification_days,
            status=status_val,
            created_by=str(new_medicine.created_by),
            created_at=new_medicine.created_at,
            updated_at=new_medicine.updated_at,
        )
    except Exception as e:
        db.rollback()
        logger.error(f"Error creating medicine: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create medicine"
        )


@router.get("/medicines/{medicine_id}", response_model=MedicineResponse)
async def get_medicine(
    medicine_id: str,
    current_user: TokenData = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Get a specific medicine"""
    medicine = db.query(Medicine).filter(
        Medicine.id == medicine_id,
        Medicine.pharmacy_id == current_user.pharmacy_id,
    ).first()
    
    if not medicine:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Medicine not found"
        )
    
    # Calculate status
    now = datetime.now().date()
    status_val = "ok"
    if medicine.expiry_date < now:
        status_val = "expired"
    elif (medicine.expiry_date - now).days <= medicine.notification_days:
        status_val = "expiring"
    elif medicine.quantity_remaining <= medicine.reorder_level:
        status_val = "low"
    
    return MedicineResponse(
        id=str(medicine.id),
        pharmacy_id=str(medicine.pharmacy_id),
        name=medicine.name,
        generic_name=medicine.generic_name,
        category=medicine.category,
        unit=medicine.unit,
        quantity_total=medicine.quantity_total,
        quantity_remaining=medicine.quantity_remaining,
        expiry_date=medicine.expiry_date.isoformat(),
        batch_number=medicine.batch_number,
        supplier=medicine.supplier,
        unit_cost=medicine.unit_cost,
        reorder_level=medicine.reorder_level,
        notification_days=medicine.notification_days,
        status=status_val,
        created_by=str(medicine.created_by),
        created_at=medicine.created_at,
        updated_at=medicine.updated_at,
    )


@router.put("/medicines/{medicine_id}", response_model=MedicineResponse)
async def update_medicine(
    medicine_id: str,
    medicine_data: MedicineUpdate,
    current_user: TokenData = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Update a medicine entry"""
    medicine = db.query(Medicine).filter(
        Medicine.id == medicine_id,
        Medicine.pharmacy_id == current_user.pharmacy_id,
    ).first()
    
    if not medicine:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Medicine not found"
        )
    
    try:
        # Update fields if provided
        if medicine_data.name:
            medicine.name = medicine_data.name
        if medicine_data.generic_name:
            medicine.generic_name = medicine_data.generic_name
        if medicine_data.category:
            medicine.category = medicine_data.category
        if medicine_data.quantity_remaining is not None:
            medicine.quantity_remaining = medicine_data.quantity_remaining
        if medicine_data.expiry_date:
            medicine.expiry_date = datetime.fromisoformat(medicine_data.expiry_date).date()
        if medicine_data.batch_number:
            medicine.batch_number = medicine_data.batch_number
        if medicine_data.supplier:
            medicine.supplier = medicine_data.supplier
        if medicine_data.unit_cost is not None:
            medicine.unit_cost = medicine_data.unit_cost
        if medicine_data.reorder_level is not None:
            medicine.reorder_level = medicine_data.reorder_level
        if medicine_data.notification_days is not None:
            medicine.notification_days = medicine_data.notification_days
        
        db.commit()
        db.refresh(medicine)
        
        # Calculate status
        now = datetime.now().date()
        status_val = "ok"
        if medicine.expiry_date < now:
            status_val = "expired"
        elif (medicine.expiry_date - now).days <= medicine.notification_days:
            status_val = "expiring"
        elif medicine.quantity_remaining <= medicine.reorder_level:
            status_val = "low"
        
        return MedicineResponse(
            id=str(medicine.id),
            pharmacy_id=str(medicine.pharmacy_id),
            name=medicine.name,
            generic_name=medicine.generic_name,
            category=medicine.category,
            unit=medicine.unit,
            quantity_total=medicine.quantity_total,
            quantity_remaining=medicine.quantity_remaining,
            expiry_date=medicine.expiry_date.isoformat(),
            batch_number=medicine.batch_number,
            supplier=medicine.supplier,
            unit_cost=medicine.unit_cost,
            reorder_level=medicine.reorder_level,
            notification_days=medicine.notification_days,
            status=status_val,
            created_by=str(medicine.created_by),
            created_at=medicine.created_at,
            updated_at=medicine.updated_at,
        )
    except Exception as e:
        db.rollback()
        logger.error(f"Error updating medicine: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update medicine"
        )


@router.delete("/medicines/{medicine_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_medicine(
    medicine_id: str,
    current_user: TokenData = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Delete a medicine entry"""
    medicine = db.query(Medicine).filter(
        Medicine.id == medicine_id,
        Medicine.pharmacy_id == current_user.pharmacy_id,
    ).first()
    
    if not medicine:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Medicine not found"
        )
    
    try:
        db.delete(medicine)
        db.commit()
    except Exception as e:
        db.rollback()
        logger.error(f"Error deleting medicine: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete medicine"
        )
